### PR TITLE
libretro: refactor infrastructure

### PR DESCRIPTION
**Pull request by thiagokokada**
This is a full refactor on how we package Libretro cores inside `nixpkgs`. The idea is to remove all our custom infrastructure (e.g.: `update_cores.py`) in favor of just having each core in their own file calling the `mkLibretroCore` function. This has lots of advantages:
- We don't need a custom update script anymore, we can simple use the existing update scripts
- Much easier to reason about each core, instead of looking at 3 different files

All cores are in their own scope (this is the reason we can't move the cores to `pkgs/by-name` though, since they have their own scope), making it much easier to override dependencies cross packages.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/emulators/libretro/cores/beetle-supergrafx.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+  lib,
+  fetchFromGitHub,
+  mkLibretroCore,
+}:
+mkLibretroCore {
+  core = "mednafen-supergrafx";
+  version = "unstable-2024-11-15";
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/libretro/cores/beetle-supergrafx.nix:8]:
```suggestion
  version = "0-unstable-2024-11-15";
```

Comment (thiagokokada) [reply pkgs/applications/emulators/libretro/cores/beetle-supergrafx.nix]: Done.

File: pkgs/applications/emulators/libretro/cores/atari800.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+  lib,
+  fetchFromGitHub,
+  mkLibretroCore,
+}:
+mkLibretroCore {
+  core = "atari800";
+  version = "0-0-unstable-2024-10-31";
```
Comment (AndersonTorres) [inline pkgs/applications/emulators/libretro/cores/atari800.nix:8]: Ops!

Comment (thiagokokada) [reply pkgs/applications/emulators/libretro/cores/atari800.nix]: Done.

Run a `rg 0-0` to make sure that we don't have any more matches.

File: pkgs/applications/emulators/libretro/cores/dosbox.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  version = "0-unstable-2022-07-18";
+
+  src = fetchFromGitHub {
+    owner = "libretro";
+    repo = "dosbox-libretro";
+    rev = "b7b24262c282c0caef2368c87323ff8c381b3102";
+    hash = "sha256-PG2eElenlEpu0U/NIh53p0uLqewnEdaq6Aoak5E1P3I=";
+  };
+
+  CXXFLAGS = "-std=gnu++11";
```
[the remark this answers is not in the document]
Comment (thiagokokada) [reply pkgs/applications/emulators/libretro/cores/dosbox.nix:17]: Also changed in `play`.

**Comment by thiagokokada** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 357228`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 9 packages built:</summary>
  <ul>
    <li>kodi-advanced-launchers</li>
    <li>kodiPackages.libretro-2048</li>
    <li>kodiPackages.libretro-fuse</li>
    <li>kodiPackages.libretro-genplus</li>
    <li>kodiPackages.libretro-mgba</li>
    <li>kodiPackages.libretro-nestopia</li>
    <li>kodiPackages.libretro-snes9x</li>
    <li>retroarch-bare</li>
    <li>retroarch-full</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
@ofborg build retroarch-full

CC @aanderse for review.

**Comment by thiagokokada** [comment]
@ofborg build libretro

**Comment by thiagokokada** [comment]
Ok, the number of rebuilt packages looks wrong, I will take a further look what is happening later.

**Comment by thiagokokada** [comment]
I think the original PR was over ambitious. I removed all changes from RetroArch and will focus this PR in just refactoring the libretro cores. Once this is merged I will do another PR refactoring the RetroArch derivations themselves.

File: pkgs/applications/emulators/retroarch/README.md [hunk 4]
```diff
@@ -1,46 +1,16 @@
...
-./pkgs/applications/emulators/retroarch/update_cores.nix
-```
-
-Keep in mind that because of the huge amount of cores that we package here, it
-is recommended to set `GITHUB_TOKEN` to your GitHub's [Personal Access
-Token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens)
-(PAT), otherwise the update will probably fail due to GitHub's API rate limit.
+1. Add a new core using `mkLibretroCore` function (use one of the existing
+   files as an example)
+2. Add your new core to [`default.nix`](./default.nix) file
```
Comment (thiagokokada) [inline pkgs/applications/emulators/retroarch/README.md:12]: This is actually wrong but will be fixed in the next PR so I will not border fixing this.

**Comment by thiagokokada** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 357228`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 7 packages built:</summary>
  <ul>
    <li>kodiPackages.libretro-2048</li>
    <li>kodiPackages.libretro-fuse</li>
    <li>kodiPackages.libretro-genplus</li>
    <li>kodiPackages.libretro-mgba</li>
    <li>kodiPackages.libretro-nestopia</li>
    <li>kodiPackages.libretro-snes9x</li>
    <li>retroarchFull</li>
  </ul>
</details>

**Comment by thiagokokada** [comment]
> 100 packages built

Finally it is working correctly.

---

Tested:
- Gambatte
- Sameboy
- 2048
- snes9x
- bsnes
- PCSX-Rearmed
- SwanStation

CC @aanderse for review.

**Review by AndersonTorres — APPROVED** [review]

**Review by aanderse — APPROVED** [review]
this looks great! thanks for your work :heart:

**Pull request merged**